app/src/main/java/ai/brokk/context/Context.java

```diff
@@ -189,75 +190,113 @@ public String getReadOnlyToc() {
                     }
                     return true;
                 })
                 .collect(Collectors.toCollection(ArrayList::new));
 
-        newFragments.addAll(toAdd);
-
-        String actionDetails =
-                toAdd.stream().map(ContextFragment::shortDescription).collect(Collectors.joining(", "));
-        String action = "Edit " + actionDetails;
-        return withFragments(newFragments, CompletableFuture.completedFuture(action));
-    }
-
-    public Context addVirtualFragments(Collection<? extends ContextFragment.VirtualFragment> toAdd) {
-        if (toAdd.isEmpty()) {
-            return this;
-        }
-
-        var newFragments = new ArrayList<>(fragments);
-        var existingVirtuals = fragments.stream()
-                .filter(f -> f.getType().isVirtual())
-                .map(f -> (ContextFragment.VirtualFragment) f)
+        // 4. Calculate the ACTUAL new items to add.
+        //    We filter 'uniqueInputs' to ensure we don't add something that already exists
+        //    in the (cleaned) existing list.
+        var fragmentsToAdd = uniqueInputs.stream()
+                .filter(input -> keptExistingFragments.stream().noneMatch(existing -> existing.hasSameSource(input)))
                 .toList();
 
-        for (var fragment : toAdd) {
-            // Deduplicate using hasSameSource for semantic equivalence
-            boolean isDuplicate = existingVirtuals.stream().anyMatch(vf -> vf.hasSameSource(fragment))
-                    || newFragments.stream()
-                            .filter(f -> f.getType().isVirtual())
-                            .map(f -> (ContextFragment.VirtualFragment) f)
-                            .anyMatch(vf -> vf.hasSameSource(fragment));
-
-            if (!isDuplicate) {
-                newFragments.add(fragment);
-            }
+        if (fragmentsToAdd.isEmpty()) {
+            // If we filtered out skeletons but didn't actually add anything new (rare but possible),
+            // we might still want to return the updated context with skeletons removed.
+            // If strict "no-op if nothing added" is preferred, return 'this';
+            // otherwise return the keptExistingFragments.
+            // Assuming we return updated context:
+            return this.withFragments(keptExistingFragments, CompletableFuture.completedFuture(null));
         }
 
-        if (newFragments.size() == fragments.size()) {
-            return this;
-        }
+        // 5. Merge and Build Action
+        keptExistingFragments.addAll(fragmentsToAdd);
+        String action = actionBuilder.apply(fragmentsToAdd);
 
-        int addedCount = newFragments.size() - fragments.size();
-        String action = "Added " + addedCount + " fragment" + (addedCount == 1 ? "" : "s");
-        return withFragments(newFragments, CompletableFuture.completedFuture(action));
+        return this.withFragments(keptExistingFragments, CompletableFuture.completedFuture(action));
```

Comment:
Should we consider putting `CompletableFuture.completedFuture(action)` into an actual async future from cm.submitBackgroundTask?